R/R/prophet.R

```diff
@@ -1274,37 +1268,203 @@ fit.prophet <- function(m, df, ...) {
+    prior.scales = out.fit$prior.scales, 
+    type = "predict"
+  )
+  
+  # Run stan
+  if (exists(".prophet.stan.model")) {
+    model <- .prophet.stan.model
+  } else {
+    model <- get_prophet_stan_model()
   }
-  if (object$logistic.floor) {
-    cols <- c(cols, 'floor')
+  
+  stan_init <- 
+    list(
+      k = m$params$k,
+      m = m$params$m,
+      delta = as.numeric(m$params$delta),
+      beta = as.numeric(m$params$beta),
+      sigma_obs = m$params$sigma_obs
+    )
+  
+  if (m$mcmc.samples > 0) {
+    args <- list(
+      object = model,
+      data = dat,
+      init = stan_init,
+      iter = m$mcmc.samples
+    )
+    args <- utils::modifyList(args, list(...))
+    stan.fit <- do.call(rstan::sampling, args)
+  } else {
+    args <- list(
+      object = model,
+      data = dat,
+      init = stan_init,
+      iter = 0,
+      as_vector = FALSE,
+      algorithm = 'Newton',
+      verbose = FALSE
+    )
+    args <- utils::modifyList(args, list(...))
+    stan.fit <- do.call(rstan::optimizing, args)
+    
+    # Extract overall and trend predictions
+    stan.predictions <- 
+      stan.fit$par[c("y_hat", "trend_hat")] %>%
+      as.data.frame() %>%
+      mutate(yhat = y_hat, trend = trend_hat)
+    
+    # Extract overall and trend intervals
+    stan.intervals <- predict_uncertainty(m, stan.fit)
   }
-  df <- df[cols]
-  df <- dplyr::bind_cols(df, seasonal.components, intervals)
-  df$yhat <- df$trend * (1 + df$multiplicative_terms) + df$additive_terms
+  
+  df <- 
+    cbind(
+      ds = df$ds, 
+      stan.predictions, 
+      stan.intervals
+    ) %>% 
+    select( # Retain only overall and trend predictions
+      ds, 
+      yhat, yhat_lower, yhat_upper, 
+      trend, trend_lower, trend_upper
+    ) %>% 
+    mutate_at( # Rescale overall and trend predictions and intervals
+      c(
+        "yhat", "yhat_lower", "yhat_upper",
+        "trend", "trend_lower", "trend_upper"
+      ),
+      funs(. * m$y.scale + df$floor)
+    ) %>% # Now add seasonal breakdowns and intervals
+    cbind(
+      predict_seasonal_components(m, df)
+    )
+  
   return(df)
 }
 
+
+#' Output a list of all data required for stan fitting or prediction
+#'
+#' @param m The fitted prophet object
+#' @param df Data frame with historical or future data for fitting or prediction resp.
+#' @param fit.seasonal.features matrix of seasonal, holiday & exogenous regressors for fitting
+#' @param predict.seasonal.features  matrix of seasonal, holiday & exogenous regressors for predictions
+#' @param component.cols indicators for additive or multiplicative components in seasonal.features
+#' @param prior.scales prior scale hyperparameters
+#' @param type 'fit' or 'predict'
+#' 
+#' @return List with all data required for stan fitting or prediction
+#'
+#' @keywords internal
+data_for_stan <- function(
+  m, df, 
+  fit.seasonal.features, 
+  predict.seasonal.features = matrix(0, 0, ncol(fit.seasonal.features)),
```

Comment:
Optional arguments (with default values) should be after non-optional arguments.